File: src/bin/ingest_bench.rs
//! Stage-level ingestion profiler (timing lives in support scripts, not the engine).
//!
//! ```text
//! cargo run --release --bin ingest-bench
//! cargo run --release --features glove --bin ingest-bench -- --dataset glove
//! cargo run --release --bin ingest-bench -- --npy data/openai-1536.npy --limit 50000
//! cargo run --release --bin ingest-bench -- --synthetic --limit 10000 --dim 200
//! ```

use std::path::PathBuf;
use std::time::Instant;

use anyhow::{bail, Context, Result};
use clap::Parser;
use ndarray::Array2;
use ndarray_npy::WriteNpyExt;
use vectorcache::datasets::npy::NpyReader;
use vectorcache::datasets::reader::{DatasetMeta, DatasetReader, DatasetSplit};
#[cfg(feature = "glove")]
use vectorcache::datasets::reader::open_dataset;
use vectorcache::datasets::DatasetKind;
use vectorcache::ingest::{IngestionEngine, TimingSummary};
use vectorcache::quantize::{l1_words_per_vector, quantize_4d_to_1bit_into};
use vectorcache::transform::{l2_normalize_in_place, padded_dim, SrhtRotation};

#[derive(Debug, Parser)]
#[command(name = "ingest-bench", about = "Profile ingestion stage hot paths")]
struct Args {
    /// Pre-extracted float32 NPY matrix
    #[arg(long, conflicts_with_all = ["dataset", "synthetic"])]
    npy: Option<PathBuf>,

    /// Dataset name (default: VECTORCACHE_DATASET env or glove)
    #[arg(long, env = "VECTORCACHE_DATASET", conflicts_with_all = ["npy", "synthetic"])]
    dataset: Option<String>,

    /// Generate synthetic data instead of reading a dataset
    #[arg(long, conflicts_with_all = ["npy", "dataset"])]
    synthetic: bool,

    /// Data directory (default: VECTORCACHE_DATA_DIR env or data)
    #[arg(long, env = "VECTORCACHE_DATA_DIR", default_value = "data")]
    data_dir: PathBuf,

    /// HDF5 split for GloVe (train or test)
    #[arg(long, default_value = "train")]
    split: String,

    /// Cap vectors profiled (default: full dataset)
    #[arg(long)]
    limit: Option<usize>,

    /// Vector dimension for --synthetic only
    #[arg(long, default_value_t = 200)]
    dim: usize,

    /// SRHT seed
    #[arg(long, default_value_t = 42)]
    seed: u64,

    /// Write synthetic NPY here when --synthetic is used
    #[arg(long, default_value = "data/.cache/bench-synthetic.npy")]
    synthetic_path: PathBuf,
}

enum BenchReader {
    Npy(NpyReader),
    #[cfg(feature = "glove")]
    Open(vectorcache::datasets::reader::OpenDatasetReader),
}

impl DatasetReader for BenchReader {
    fn meta(&self) -> DatasetMeta {
        match self {
            Self::Npy(r) => r.meta(),
            #[cfg(feature = "glove")]
            Self::Open(r) => r.meta(),
        }
    }

    fn next_vector_into(&mut self, out: &mut [f32]) -> Result<bool> {
        match self {
            Self::Npy(r) => r.next_vector_into(out),
            #[cfg(feature = "glove")]
            Self::Open(r) => r.next_vector_into(out),
        }
    }
}

struct LimitedReader<R> {
    inner: R,
    remaining: usize,
}

impl<R: DatasetReader> DatasetReader for LimitedReader<R> {
    fn meta(&self) -> DatasetMeta {
        self.inner.meta()
    }

    fn next_vector_into(&mut self, out: &mut [f32]) -> Result<bool> {
        if self.remaining == 0 {
            return Ok(false);
        }
        if !self.inner.next_vector_into(out)? {
            bail!("reader exhausted before reaching ingest limit");
        }
        self.remaining -= 1;
        Ok(true)
    }
}

#[derive(Debug, Default, Clone)]
struct StageTotals {
    read_ns: u64,
    normalize_ns: u64,
    srht_ns: u64,
    quantize_ns: u64,
    store_ns: u64,
    batch_copy_ns: u64,
    vectors: u64,
}

impl StageTotals {
    fn total_ns(&self) -> u64 {
        self.read_ns + self.normalize_ns + self.srht_ns + self.quantize_ns + self.store_ns + self.batch_copy_ns
    }
}

fn main() -> Result<()> {
    let args = Args::parse();
    if args.synthetic && args.limit == Some(0) {
        bail!("--limit must be > 0");
    }

    let (mut reader, source_label) = open_reader(&args)?;
    let meta = reader.meta();
    let limit = args.limit.unwrap_or(meta.count).min(meta.count);
    if limit == 0 {
        bail!("dataset has no vectors to profile");
    }

    let padded = padded_dim(meta.dim);

    println!(
        "Ingest bench: {} (dim={}, padded={}, vectors={})",
        source_label, meta.dim, padded, limit
    );
    if args.limit.is_some() && args.limit.unwrap() < meta.count {
        println!("  (capped from {} vectors in dataset)", meta.count);
    }
    println!();

    let stages = profile_stages_inner(&mut reader, meta.dim, padded, args.seed, limit)?;
    print_stage_report("Per-stage (sequential micro-profile)", &stages);

    let (reader2, _) = open_reader(&args)?;
    let wall_ns = profile_engine_inner(reader2, meta.dim, args.seed, limit)?;
    print_wall_report(wall_ns, limit, &stages);

    Ok(())
}

fn open_reader(args: &Args) -> Result<(BenchReader, String)> {
    if let Some(path) = &args.npy {
        let reader = NpyReader::open(path, "bench")?;
        let label = format!("{} ({})", path.display(), reader.meta().label);
        return Ok((BenchReader::Npy(reader), label));
    }

    if args.synthetic {
        let count = args.limit.unwrap_or(10_000);
        if count == 0 {
            bail!("--limit must be > 0 for --synthetic");
        }
        ensure_synthetic_npy(&args.synthetic_path, count, args.dim)?;
        let reader = NpyReader::open(&args.synthetic_path, "synthetic")?;
        return Ok((
            BenchReader::Npy(reader),
            format!("{} (synthetic)", args.synthetic_path.display()),
        ));
    }

    let name = args.dataset.as_deref().unwrap_or("glove");
    let kind = DatasetKind::parse(name).with_context(|| format!("unknown dataset '{name}'"))?;
    let split = match args.split.as_str() {
        "train" => DatasetSplit::Train,
        "test" => DatasetSplit::Test,
        other => bail!("unknown split '{other}'; use train or test"),
    };

    #[cfg(feature = "glove")]
    {
        let reader = open_dataset(kind, &args.data_dir, split)?;
        let label = format!("{name} ({})", args.data_dir.display());
        Ok((BenchReader::Open(reader), label))
    }

    #[cfg(not(feature = "glove"))]
    {
        let _ = (kind, split);
        if matches!(kind, DatasetKind::OpenAi1536 | DatasetKind::OpenAi3072) {
            let path = kind.path(&args.data_dir);
            let reader = NpyReader::open(&path, kind.label())?;
            let label = format!("{name} ({})", path.display());
            return Ok((BenchReader::Npy(reader), label));
        }
        bail!(
            "dataset '{name}' requires --npy, --synthetic, or build with --features glove"
        );
    }
}

fn ensure_synthetic_npy(path: &PathBuf, count: usize, dim: usize) -> Result<()> {
    if path.is_file() {
        let reader = NpyReader::open(path, "synthetic")?;
        if reader.meta().dim == dim && reader.meta().count >= count {
            return Ok(());
        }
    }

    if let Some(parent) = path.parent() {
        std::fs::create_dir_all(parent)?;
    }

    let data: Vec<f32> = (0..count * dim)
        .map(|i| (i as f32 * 0.001313).sin() * 0.5 + (i as f32 * 0.0007).cos() * 0.3)
        .collect();
    let array = Array2::from_shape_vec((count, dim), data).context("invalid synthetic shape")?;
    let mut file = std::fs::File::create(path)
        .with_context(|| format!("failed to create {}", path.display()))?;
    array
        .write_npy(&mut file)
        .with_context(|| format!("failed to write {}", path.display()))?;
    println!("Wrote synthetic NPY: {} ({count} x {dim})", path.display());
    Ok(())
}

fn profile_stages_inner<R: DatasetReader>(
    reader: &mut R,
    dim: usize,
    padded: usize,
    seed: u64,
    limit: usize,
) -> Result<StageTotals> {
    use vectorcache::ingest::INGEST_BATCH_SIZE;

    let rotation = SrhtRotation::new(dim, seed);
    let l1_words = l1_words_per_vector(padded);
    let batch_cap = INGEST_BATCH_SIZE.min(limit.max(1));

    let mut read_buf = vec![0.0; dim];
    let mut batch_inputs = vec![0.0_f32; batch_cap * dim];
    let mut batch_work: Vec<(Vec<f32>, Vec<f32>, Vec<u64>)> = (0..batch_cap)
        .map(|_| (vec![0.0; dim], vec![0.0; padded], vec![0_u64; l1_words]))
        .collect();

    let mut store = vectorcache::ingest::BlockStore::with_capacity(l1_words, limit);
    let mut totals = StageTotals::default();
    let mut processed = 0usize;

    while processed < limit {
        let mut batch_len = 0usize;
        while batch_len < batch_cap && processed + batch_len < limit {
            let t0 = Instant::now();
            if !reader.next_vector_into(&mut read_buf)? {
                bail!("reader exhausted at {} vectors", processed + batch_len);
            }
            totals.read_ns += t0.elapsed().as_nanos() as u64;

            let dst = batch_len * dim;
            batch_inputs[dst..dst + dim].copy_from_slice(&read_buf);
            batch_len += 1;
        }
        if batch_len == 0 {
            break;
        }

        let inputs = {
            let t_copy = Instant::now();
            let copied = batch_inputs[..batch_len * dim].to_vec();
            totals.batch_copy_ns += t_copy.elapsed().as_nanos() as u64;
            copied
        };

        for i in 0..batch_len {
            let (normalized, rotated, l1) = &mut batch_work[i];
            normalized.copy_from_slice(&inputs[i * dim..(i + 1) * dim]);

            let t_norm = Instant::now();
            l2_normalize_in_place(normalized);
            totals.normalize_ns += t_norm.elapsed().as_nanos() as u64;

            let t_srht = Instant::now();
            rotation.apply(normalized, rotated)?;
            totals.srht_ns += t_srht.elapsed().as_nanos() as u64;

            let t_quant = Instant::now();
            quantize_4d_to_1bit_into(rotated, l1);
            totals.quantize_ns += t_quant.elapsed().as_nanos() as u64;

            let t_store = Instant::now();
            store.push_l1_codes(l1)?;
            totals.store_ns += t_store.elapsed().as_nanos() as u64;
        }

        processed += batch_len;
        totals.vectors = processed as u64;
    }

    Ok(totals)
}

fn profile_engine_inner<R: DatasetReader>(
    reader: R,
    dim: usize,
    seed: u64,
    limit: usize,
) -> Result<u64> {
    let mut limited = LimitedReader {
        inner: reader,
        remaining: limit,
    };
    let mut engine = IngestionEngine::with_rotation(dim, seed);
    engine.reserve_vectors(limit);

    let start = Instant::now();
    let report = engine.ingest(&mut limited)?;
    let wall_ns = start.elapsed().as_nanos() as u64;

    if report.vectors_ingested as usize != limit {
        bail!(
            "engine ingested {} vectors, expected {}",
            report.vectors_ingested,
            limit
        );
    }

    Ok(wall_ns)
}

fn print_stage_report(label: &str, stages: &StageTotals) {
    let total = stages.total_ns().max(1);
    let v = stages.vectors.max(1) as f64;

    println!("{label}");
    println!("  vectors: {}", stages.vectors);
    println!(
        "  total:   {} ({:.0} ns)",
        TimingSummary::format_duration(total),
        total
    );
    println!();

    let rows = [
        ("read (mmap → buffer)", stages.read_ns),
        ("batch copy (engine-style to_vec)", stages.batch_copy_ns),
        ("L2 normalize", stages.normalize_ns),
        ("SRHT (3× sign + FWHT)", stages.srht_ns),
        ("L1 quantize", stages.quantize_ns),
        ("store (push_l1_codes)", stages.store_ns),
    ];

    println!("  {:<34} {:>12} {:>10} {:>7}", "stage", "total", "per-vec", "share");
    println!("  {}", "-".repeat(67));
    for (name, ns) in rows {
        let pct = ns as f64 / total as f64 * 100.0;
        println!(
            "  {:<34} {:>12} {:>10} {:>6.1}%",
            name,
            TimingSummary::format_duration(ns),
            TimingSummary::format_duration((ns as f64 / v) as u64),
            pct
        );
    }
    println!();
}

fn print_wall_report(wall_ns: u64, vectors: usize, stages: &StageTotals) {
    let v = vectors.max(1) as f64;
    let seq_total = stages.total_ns();

    println!("Engine ingest (parallel batches, release)");
    println!(
        "  wall:    {} ({:.0} ns)",
        TimingSummary::format_duration(wall_ns),
        wall_ns
    );
    println!(
        "  per-vec: {} ({:.0} ns)",
        TimingSummary::format_duration((wall_ns as f64 / v) as u64),
        wall_ns as f64 / v
    );
    println!();

    if wall_ns > 0 {
        let parallel_speedup = seq_total as f64 / wall_ns as f64;
        println!(
            "  parallel speedup vs sequential stages: {:.2}x",
            parallel_speedup
        );
    }

    println!();
    println!("Hot paths (by sequential stage share):");
    let mut ranked = [
        ("SRHT / FWHT", stages.srht_ns),
        ("batch input copy (to_vec per batch)", stages.batch_copy_ns),
        ("L1 quantize", stages.quantize_ns),
        ("read I/O", stages.read_ns),
        ("L2 normalize", stages.normalize_ns),
        ("store", stages.store_ns),
    ];
    ranked.sort_by(|a, b| b.1.cmp(&a.1));
    for (i, (name, ns)) in ranked.iter().enumerate() {
        let pct = *ns as f64 / seq_total.max(1) as f64 * 100.0;
        println!("  {}. {} — {:.1}%", i + 1, name, pct);
    }
}

File: tools/ingest-bench.cpp
#include <algorithm>
#include <chrono>
#include <cstring>
#include <iomanip>
#include <iostream>
#include <memory>
#include <optional>
#include <string>
#include <vector>

#include <CLI/CLI.hpp>

#include "vectorcache/datasets/datasets.hpp"
#include "vectorcache/datasets/npy.hpp"
#include "vectorcache/error.hpp"
#include "vectorcache/ingest/engine.hpp"
#include "vectorcache/ingest/store.hpp"
#include "vectorcache/ingest/timing.hpp"
#include "vectorcache/quantize/quantize.hpp"
#include "vectorcache/transform/fwht.hpp"
#include "vectorcache/transform/normalize.hpp"
#include "vectorcache/transform/srht.hpp"
#include "vectorcache/transform/srht_config.hpp"

namespace {

#if VECTORCACHE_SRHT_ROUNDS == 1
constexpr const char* kSrhtStageLabel = "SRHT (1x sign + FWHT)";
#elif VECTORCACHE_SRHT_ROUNDS == 2
constexpr const char* kSrhtStageLabel = "SRHT (2x sign + FWHT)";
#else
constexpr const char* kSrhtStageLabel = "SRHT (3x sign + FWHT)";
#endif

struct StageTotals {
  std::uint64_t read_ns = 0;
  std::uint64_t normalize_ns = 0;
  std::uint64_t srht_ns = 0;
  std::uint64_t quantize_ns = 0;
  std::uint64_t store_ns = 0;
  std::uint64_t batch_copy_ns = 0;
  std::uint64_t vectors = 0;

  std::uint64_t total_ns() const {
    return read_ns + normalize_ns + srht_ns + quantize_ns + store_ns + batch_copy_ns;
  }
};

class LimitedReader : public vectorcache::datasets::DatasetReader {
 public:
  LimitedReader(vectorcache::datasets::DatasetReader& inner, std::size_t remaining)
      : inner_(inner), remaining_(remaining) {}

  vectorcache::datasets::DatasetMeta meta() const override { return inner_.meta(); }

  bool next_vector_into(std::span<float> out) override {
    if (remaining_ == 0) return false;
    if (!inner_.next_vector_into(out)) {
      throw vectorcache::Error("reader exhausted before reaching ingest limit");
    }
    --remaining_;
    return true;
  }

 private:
  vectorcache::datasets::DatasetReader& inner_;
  std::size_t remaining_;
};

std::pair<std::unique_ptr<vectorcache::datasets::DatasetReader>, std::string> open_reader(
    const std::filesystem::path& npy_path, const std::string& dataset,
    const std::filesystem::path& data_dir, const std::string& split) {
  if (!npy_path.empty()) {
    auto reader = std::make_unique<vectorcache::datasets::NpyReader>(
        vectorcache::datasets::NpyReader::open(npy_path, "bench"));
    return {std::move(reader), npy_path.string() + " (" + reader->meta().label + ")"};
  }

  if (dataset.empty()) {
    throw vectorcache::Error("--dataset is required (or pass --npy)");
  }

  auto kind = vectorcache::datasets::parse_dataset_kind(dataset);
  if (!kind) throw vectorcache::Error("unknown dataset '" + dataset + "'");

  vectorcache::datasets::DatasetSplit dataset_split = vectorcache::datasets::DatasetSplit::Train;
  if (split == "test") {
    dataset_split = vectorcache::datasets::DatasetSplit::Test;
  } else if (split != "train") {
    throw vectorcache::Error("unknown split '" + split + "'");
  }

#ifdef VECTORCACHE_BUILD_GLOVE
  auto reader = std::make_unique<vectorcache::datasets::OpenDatasetReader>(
      vectorcache::datasets::OpenDatasetReader::open(*kind, data_dir, dataset_split));
  return {std::move(reader), dataset + " (" + data_dir.string() + ")"};
#else
  if (*kind == vectorcache::datasets::DatasetKind::OpenAi1536 ||
      *kind == vectorcache::datasets::DatasetKind::OpenAi3072) {
    const auto path = vectorcache::datasets::dataset_path(*kind, data_dir);
    auto reader = std::make_unique<vectorcache::datasets::NpyReader>(
        vectorcache::datasets::NpyReader::open(path, vectorcache::datasets::dataset_label(*kind)));
    return {std::move(reader), dataset + " (" + path.string() + ")"};
  }
  throw vectorcache::Error("dataset requires --npy or VECTORCACHE_BUILD_GLOVE");
#endif
}

StageTotals profile_stages(vectorcache::datasets::DatasetReader& reader, std::size_t dim,
                           std::size_t padded, std::uint64_t seed, std::size_t limit) {
  const vectorcache::transform::SrhtRotation rotation(dim, seed);
  const std::size_t l1_words = vectorcache::quantize::l1_words_per_vector(padded);
  const std::size_t l0_words = vectorcache::quantize::l0_words_per_vector(padded);
  const std::size_t batch_cap =
      std::min(vectorcache::ingest::INGEST_BATCH_SIZE, std::max(limit, std::size_t{1}));

  std::vector<float> read_buf(dim);
  std::vector<float> batch_inputs(batch_cap * dim);
  std::vector<std::vector<float>> rotated(batch_cap, std::vector<float>(padded));
  std::vector<std::vector<std::uint64_t>> l1(batch_cap, std::vector<std::uint64_t>(l1_words));
  std::vector<std::vector<std::uint64_t>> l0(batch_cap, std::vector<std::uint64_t>(l0_words));

  auto store =
      vectorcache::ingest::BlockStore::with_capacity(l1_words, l0_words, padded, limit);
  StageTotals totals;
  std::size_t processed = 0;

  while (processed < limit) {
    std::size_t batch_len = 0;
    while (batch_len < batch_cap && processed + batch_len < limit) {
      const auto t0 = std::chrono::steady_clock::now();
      if (!reader.next_vector_into(read_buf)) {
        throw vectorcache::Error("reader exhausted during profiling");
      }
      totals.read_ns += static_cast<std::uint64_t>(
          std::chrono::duration_cast<std::chrono::nanoseconds>(std::chrono::steady_clock::now() -
                                                               t0)
              .count());
      std::copy(read_buf.begin(), read_buf.end(),
                batch_inputs.begin() + static_cast<std::ptrdiff_t>(batch_len * dim));
      ++batch_len;
    }
    if (batch_len == 0) break;

    for (std::size_t i = 0; i < batch_len; ++i) {
      const float* input =
          batch_inputs.data() + static_cast<std::ptrdiff_t>(i * dim);

      std::memcpy(rotated[i].data(), input, dim * sizeof(float));
      if (padded > dim) {
        std::memset(rotated[i].data() + dim, 0, (padded - dim) * sizeof(float));
      }

      const auto t_norm = std::chrono::steady_clock::now();
      vectorcache::transform::l2_normalize_in_place(rotated[i]);
      totals.normalize_ns += static_cast<std::uint64_t>(
          std::chrono::duration_cast<std::chrono::nanoseconds>(std::chrono::steady_clock::now() -
                                                               t_norm)
              .count());

      const auto t_srht = std::chrono::steady_clock::now();
      rotation.apply_in_place(rotated[i]);
      totals.srht_ns += static_cast<std::uint64_t>(
          std::chrono::duration_cast<std::chrono::nanoseconds>(std::chrono::steady_clock::now() -
                                                               t_srht)
              .count());

      const auto t_quant = std::chrono::steady_clock::now();
      vectorcache::quantize::quantize_4d_to_1bit_into(rotated[i], l1[i]);
      vectorcache::quantize::quantize_1dim_to_1bit_into(rotated[i], l0[i]);
      totals.quantize_ns += static_cast<std::uint64_t>(
          std::chrono::duration_cast<std::chrono::nanoseconds>(std::chrono::steady_clock::now() -
                                                               t_quant)
              .count());

      const auto t_store = std::chrono::steady_clock::now();
      store.push_vector(l1[i], l0[i]);
      totals.store_ns += static_cast<std::uint64_t>(
          std::chrono::duration_cast<std::chrono::nanoseconds>(std::chrono::steady_clock::now() -
                                                               t_store)
              .count());
    }

    processed += batch_len;
    totals.vectors = processed;
  }

  return totals;
}

std::uint64_t profile_engine(vectorcache::datasets::DatasetReader& reader, std::size_t dim,
                             std::uint64_t seed, std::size_t limit) {
  LimitedReader limited(reader, limit);
  auto engine = vectorcache::ingest::IngestionEngine::with_rotation(dim, seed);
  engine.reserve_vectors(limit);
  const auto start = std::chrono::steady_clock::now();
  const auto report = engine.ingest(limited);
  const auto wall_ns = static_cast<std::uint64_t>(
      std::chrono::duration_cast<std::chrono::nanoseconds>(std::chrono::steady_clock::now() -
                                                           start)
          .count());
  if (report.vectors_ingested != limit) {
    throw vectorcache::Error("engine ingested unexpected vector count");
  }
  return wall_ns;
}

void print_stage_report(const std::string& label, const StageTotals& stages) {
  const std::uint64_t total = std::max<std::uint64_t>(stages.total_ns(), 1);
  const double v = std::max<std::uint64_t>(stages.vectors, 1);

  std::cout << label << '\n';
  std::cout << "  vectors: " << stages.vectors << '\n';
  std::cout << "  total:   " << vectorcache::ingest::TimingSummary::format_duration(total) << " ("
            << total << " ns)\n\n";

  const std::pair<const char*, std::uint64_t> rows[] = {
      {"read (mmap -> buffer)", stages.read_ns},
      {"batch copy (engine-style to_vec)", stages.batch_copy_ns},
      {"L2 normalize", stages.normalize_ns},
      {kSrhtStageLabel, stages.srht_ns},
      {"L1+L0 quantize", stages.quantize_ns},
      {"store (push_vector)", stages.store_ns},
  };

  std::cout << "  " << std::left << std::setw(34) << "stage" << std::right << std::setw(12)
            << "total" << std::setw(10) << "per-vec" << std::setw(7) << "share\n";
  std::cout << "  " << std::string(67, '-') << '\n';
  for (const auto& [name, ns] : rows) {
    const double pct = static_cast<double>(ns) / static_cast<double>(total) * 100.0;
    std::cout << "  " << std::left << std::setw(34) << name << std::right << std::setw(12)
              << vectorcache::ingest::TimingSummary::format_duration(ns) << std::setw(10)
              << vectorcache::ingest::TimingSummary::format_duration(
                     static_cast<std::uint64_t>(static_cast<double>(ns) / v))
              << std::setw(6) << std::fixed << std::setprecision(1) << pct << "%\n";
  }
  std::cout << '\n';
}

void print_wall_report(std::uint64_t wall_ns, std::size_t vectors, const StageTotals& stages) {
  const double v = std::max(vectors, std::size_t{1});
  const std::uint64_t seq_total = stages.total_ns();

  std::cout << "Engine ingest (release)\n";
  std::cout << "  wall:    " << vectorcache::ingest::TimingSummary::format_duration(wall_ns)
            << " (" << wall_ns << " ns)\n";
  std::cout << "  per-vec: "
            << vectorcache::ingest::TimingSummary::format_duration(
                   static_cast<std::uint64_t>(static_cast<double>(wall_ns) / v))
            << " (" << (static_cast<double>(wall_ns) / v) << " ns)\n\n";

  if (seq_total > 0) {
    const double overhead_pct =
        (static_cast<double>(wall_ns) - static_cast<double>(seq_total)) /
        static_cast<double>(seq_total) * 100.0;
    std::cout << "  engine overhead vs sequential stages: " << std::fixed << std::setprecision(1)
              << overhead_pct << "%\n";
  }

  std::cout << "\nHot paths (by sequential stage share):\n";
  std::vector<std::pair<const char*, std::uint64_t>> ranked = {
      {"SRHT / FWHT", stages.srht_ns},
      {"batch input copy (to_vec per batch)", stages.batch_copy_ns},
      {"L1+L0 quantize", stages.quantize_ns},
      {"read I/O", stages.read_ns},
      {"L2 normalize", stages.normalize_ns},
      {"store", stages.store_ns},
  };
  std::sort(ranked.begin(), ranked.end(),
            [](const auto& a, const auto& b) { return a.second > b.second; });
  for (std::size_t i = 0; i < ranked.size(); ++i) {
    const double pct =
        static_cast<double>(ranked[i].second) /
        static_cast<double>(std::max(seq_total, std::uint64_t{1})) *
        100.0;
    std::cout << "  " << (i + 1) << ". " << ranked[i].first << " — " << std::fixed
              << std::setprecision(1) << pct << "%\n";
  }
}

}  // namespace

int main(int argc, char** argv) {
  CLI::App app{"Profile ingestion stage hot paths"};
  std::filesystem::path npy_path;
  std::string dataset;
  std::filesystem::path data_dir = "data";
  std::string split = "train";
  std::optional<std::size_t> limit;
  std::uint64_t seed = 42;

  app.add_option("--npy", npy_path, "Pre-extracted float32 NPY matrix");
  app.add_option("--dataset", dataset, "Dataset name")->envname("VECTORCACHE_DATASET");
  app.add_option("--data-dir", data_dir, "Data directory")->envname("VECTORCACHE_DATA_DIR");
  app.add_option("--split", split, "HDF5 split for GloVe");
  app.add_option("--limit", limit, "Cap vectors profiled");
  app.add_option("--seed", seed, "SRHT seed");

  CLI11_PARSE(app, argc, argv);

  try {
    if (npy_path.empty() && dataset.empty()) {
      throw vectorcache::Error("pass --dataset or --npy");
    }

    auto [reader1, source_label] = open_reader(npy_path, dataset, data_dir, split);
    const auto meta = reader1->meta();
    const std::size_t actual_limit = std::min(limit.value_or(meta.count), meta.count);
    if (actual_limit == 0) {
      throw vectorcache::Error("dataset has no vectors to profile");
    }

    const std::size_t padded = vectorcache::transform::padded_dim(meta.dim);
    std::cout << "Ingest bench: " << source_label << " (dim=" << meta.dim << ", padded=" << padded
              << ", vectors=" << actual_limit
              << ", srht_rounds=" << vectorcache::transform::srht_rounds() << ")\n";
    if (limit && *limit < meta.count) {
      std::cout << "  (capped from " << meta.count << " vectors in dataset)\n";
    }
    std::cout << '\n';

    const auto stages = profile_stages(*reader1, meta.dim, padded, seed, actual_limit);
    print_stage_report("Per-stage (sequential micro-profile)", stages);

    auto [reader2, _] = open_reader(npy_path, dataset, data_dir, split);
    const auto wall_ns = profile_engine(*reader2, meta.dim, seed, actual_limit);
    print_wall_report(wall_ns, actual_limit, stages);

    return 0;
  } catch (const vectorcache::Error& e) {
    std::cerr << e.what() << '\n';
    return 1;
  }
}
